python-package/lightgbm/basic.py

```diff
@@ -1040,8 +1042,9 @@ def subset(self, used_indices, params=None):
         """
         if params is None:
             params = self.params
-        ret = Dataset(None, reference=self, feature_name=self.feature_name,
-                      categorical_feature=self.categorical_feature, params=params)
+        ret = Dataset(self.data, reference=self, feature_name=self.feature_name,
+                      categorical_feature=self.categorical_feature, params=params,
+                      free_raw_data=self.free_raw_data)
```

Comment:
maybe a better logic still pass `None` for `data` here?
Then in `construct`, use reference.data to `get_data` ?
Then, the logic of construct are the same as before.